include/grpcpp/impl/codegen/server_callback.h

```diff
@@ -60,22 +60,19 @@ class ServerReactor {
   template <class Request, class Response>
   friend class CallbackBidiHandler;
 
-  enum class CancellationState {
-    ON_STARTED_NOT_YET_DONE,
-    ON_STARTED_DONE,
-    ON_CANCEL_NEEDED
-  };
-
-  bool MaybeChangeCancellationState(CancellationState new_state,
-                                    std::memory_order success_order,
-                                    std::memory_order failure_order) {
-    CancellationState state{CancellationState::ON_STARTED_NOT_YET_DONE};
-    return cancellation_state_.compare_exchange_strong(
-        state, new_state, success_order, failure_order);
+  // The ServerReactor is responsible for tracking when it is safe to call
+  // OnCancel. This function should not be called until after OnStarted is done
+  // and the RPC has completed with a cancellation. This is tracked by counting
+  // how many of these conditions have been met and calling OnCancel when none
+  // remain unmet.
+
+  void MaybeCallOnCancel() {
+    if (--on_cancel_conditions_remaining_ == 0) {
```

Comment:
nit: To make sure future readers do not confuse this with a regular integer, I would suggest using fetch_sub(-1, memory_ordre_acq_rel).

Reply:
Done, but fetch_sub(1, ...)

Reply:
Looks great. thanks!